File: modules/iotbroker/sofia2-iotbroker-boot/src/test/java/com/indracompany/sofia2/iotbroker/mock/ssap/SSAPMessageGenerator.java
/**
 * Copyright Indra Sistemas, S.A.
 * 2013-2018 SPAIN
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.indracompany.sofia2.iotbroker.mock.ssap;

import java.io.IOException;
import java.util.UUID;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.github.javafaker.Faker;
import com.indracompany.sofia2.ssap.SSAPMessage;
import com.indracompany.sofia2.ssap.body.SSAPBodyDeleteByIdMessage;
import com.indracompany.sofia2.ssap.body.SSAPBodyDeleteMessage;
import com.indracompany.sofia2.ssap.body.SSAPBodyInsertMessage;
import com.indracompany.sofia2.ssap.body.SSAPBodyJoinMessage;
import com.indracompany.sofia2.ssap.body.SSAPBodyUpdateByIdMessage;
import com.indracompany.sofia2.ssap.body.SSAPBodyUpdateMessage;
import com.indracompany.sofia2.ssap.enums.SSAPMessageDirection;
import com.indracompany.sofia2.ssap.enums.SSAPMessageTypes;

public final class SSAPMessageGenerator {

	private static final Faker faker = new Faker();
	private static final ObjectMapper mapper = new ObjectMapper();

	public static SSAPMessage<SSAPBodyJoinMessage> generateJoinMessageWithToken() {
		final SSAPMessage<SSAPBodyJoinMessage> ssapMessage = new SSAPMessage<>();
		final SSAPBodyJoinMessage body = new SSAPBodyJoinMessage();
		ssapMessage.setDirection(SSAPMessageDirection.REQUEST);
		ssapMessage.setMessageId(UUID.randomUUID().toString());
		ssapMessage.setMessageType(SSAPMessageTypes.JOIN);
		body.setClientPlatform(faker.name().firstName());
		body.setClientPlatformInstance(UUID.randomUUID().toString());
		body.setToken(UUID.randomUUID().toString());

		ssapMessage.setBody(body);

		return ssapMessage;
	}

	public static SSAPMessage<SSAPBodyInsertMessage> generateInsertMessage(String ontology, Object value) throws Exception, IOException {

		final SSAPMessage<SSAPBodyInsertMessage> message = new SSAPMessage<>();
		message.setSessionKey(UUID.randomUUID().toString());

		final SSAPBodyInsertMessage body = new SSAPBodyInsertMessage();
		final JsonNode jsonValue = mapper.readTree(mapper.writeValueAsBytes(value));
		body.setData(jsonValue);
		body.setOntology(ontology);
		message.setBody(body);
		message.setDirection(SSAPMessageDirection.REQUEST);
		message.setMessageType(SSAPMessageTypes.INSERT);
		return message;
	}

	public static SSAPMessage<SSAPBodyUpdateMessage> generateUpdatetMessage(String ontology, String query) throws Exception, IOException {

		final SSAPMessage<SSAPBodyUpdateMessage> message = new SSAPMessage<>();
		message.setSessionKey(UUID.randomUUID().toString());

		final SSAPBodyUpdateMessage body = new SSAPBodyUpdateMessage();
		body.setQuery(query);
		body.setOntology(ontology);
		message.setBody(body);
		message.setDirection(SSAPMessageDirection.REQUEST);
		message.setMessageType(SSAPMessageTypes.UPDATE);

		return message;
	}

	public static SSAPMessage<SSAPBodyUpdateByIdMessage> generateUpdateByIdtMessage(String ontology, JsonNode data) {
		final SSAPMessage<SSAPBodyUpdateByIdMessage> message = new SSAPMessage<>();
		message.setSessionKey(UUID.randomUUID().toString());

		final SSAPBodyUpdateByIdMessage body = new SSAPBodyUpdateByIdMessage();
		body.setData(data);
		body.setOntology(ontology);
		message.setBody(body);
		message.setDirection(SSAPMessageDirection.REQUEST);
		message.setMessageType(SSAPMessageTypes.UPDATE_BY_ID);

		return message;
	}

	public static SSAPMessage<SSAPBodyDeleteMessage> generateDeleteMessage(String ontology, String query) {
		final SSAPMessage<SSAPBodyDeleteMessage> message = new SSAPMessage<>();
		message.setSessionKey(UUID.randomUUID().toString());

		final SSAPBodyDeleteMessage body = new SSAPBodyDeleteMessage();
		body.setOntology(ontology);
		body.setQuery(query);
		message.setBody(body);
		message.setDirection(SSAPMessageDirection.REQUEST);
		message.setMessageType(SSAPMessageTypes.DELETE);

		return message;
	}

	public static SSAPMessage<SSAPBodyDeleteByIdMessage> generateDeleteByIdMessage(String ontology, String id) {
		final SSAPMessage<SSAPBodyDeleteByIdMessage> message = new SSAPMessage<>();
		message.setSessionKey(UUID.randomUUID().toString());

		final SSAPBodyDeleteByIdMessage body = new SSAPBodyDeleteByIdMessage();
		body.setOntology(ontology);
		body.setId(id);
		message.setBody(body);
		message.setDirection(SSAPMessageDirection.REQUEST);
		message.setMessageType(SSAPMessageTypes.DELETE_BY_ID);

		return message;
	}


}

File: client-libraries/node-red/node-red-contrib-sofia4cities/libraries/SSAPMessageGenerator.js

function addQuotesToData(data){
	if (data.indexOf("{")!=0)
		data="{"+data+"}";

	return data;
}

var SSAPMessageGenerator = {

	/**
	 * JOIN By Token
	 */
	generateJoinByTokenMessage : function(token, device, instance) {
		var queryJoin = '{"messageId":"","sessionKey":null,"direction":"REQUEST","messageType":"JOIN","body":{"@type":"SSAPBodyJoinMessage","token":"'
		+ token + '","clientPlatform":"'
		+ device + '","clientPlatformInstance":"'
		+ instance + '"}}';

		return queryJoin;
	},

	/**
	 * JOIN Renovate By Token
	 */
	generateJoinRenovateByTokenMessage : function(token, device, instance, sessionKey) {
		var queryJoin = '{"messageId":"","sessionKey":"'
		+ sessionKey + '","direction":"REQUEST","messageType":"JOIN","body":{"@type":"SSAPBodyJoinMessage","token":"'
		+ token + '","clientPlatform":"'
		+ device + '","clientPlatformInstance":"'
		+ instance + '"}}';

		return queryJoin;
	},

	/**
	 * LEAVE Operation
	 */
	generateLeaveMessage : function(sessionKey) {
		var queryLeave = '{"messageId":"","sessionKey":"'
		+ sessionKey +'","direction":"REQUEST","messageType":"LEAVE","body":{"@type":"SSAPBodyJoinMessage","token":"null","clientPlatform":"null","clientPlatformInstance":"null"}}';

		return queryLeave;
	},

	/**
	 * INSERT Message
	 */
	generateInsertMessage : function(data, ontology, sessionKey) {
		if(ontology===undefined || ontology==null){
			ontology='null'
		}else{
			ontology='"'+ontology+'"';
		}
		var queryInsert = '{"messageId": null,"sessionKey": "'
				+ sessionKey + '","direction":"REQUEST","messageType":"INSERT","body": {"@type":"SSAPBodyInsertMessage","ontology":'
				+ ontology + ',"data":'
				+ data + '}}';

		return queryInsert;
	},

	/**
	 * QUERY with queryType Operation
	 */
	generateQueryWithQueryTypeMessage : function(query, ontology, queryType, sessionKey) {
		if(ontology===undefined || ontology==null){
			ontology='null'
		}else{
			ontology='"'+ontology+'"';
		}
		var querySib='';

		var querySib = '{"messageId": null,"sessionKey": "'
			+ sessionKey	+'","direction": "REQUEST","messageType": "QUERY","body": {"@type":"SSAPBodyQueryMessage","ontology":'
			+ ontology +',"query":"'
			+ query +'","queryType":"'
			+ queryType +'","resultFormat":null}}';

		return querySib
	},

	/**
	 * REMOVE Operation
	 */
	generateDeleteMessage : function(query, ontology, sessionKey) {
		if(ontology===undefined || ontology==null){
			ontology='null'
		}
		var queryRemove = '{"messageId": null,"sessionKey":"'
		  + sessionKey + '","direction":"REQUEST","messageType": "DELETE","body": {"@type": "SSAPBodyDeleteMessage","ontology":"'
		  + ontology + '","query":"db.'
			+ ontology + '.remove('
			+ query + ')"}}';

		return queryRemove;
	},

	/**
	 * REMOVE_BY_ID Operation
	 */
	generateDeleteByIdMessage : function(id, ontology, sessionKey) {
		if(ontology===undefined || ontology==null){
			ontology='null'
		}else{
			ontology='"'+ontology+'"';
		}
		var queryRemove = '{"messageId": null,"sessionKey":"'
		  + sessionKey + '","direction":"REQUEST","messageType": "DELETE_BY_ID","body": {"@type": "SSAPBodyDeleteByIdMessage","ontology":'
		  + ontology + ',"id": "'
			+ id + '"}}';

		return queryRemove;
	},

	/**
	 * UPDATE Operation
	 */
	generateUpdateMessage : function(query, ontology, sessionKey) {
		if(ontology===undefined || ontology==null){
			ontology='null'
		}else{
			ontology='"'+ontology+'"';
		}

		var queryUpdate = '{"messageId": null,"sessionKey":"'
			+ sessionKey + '","direction": "REQUEST","messageType": "UPDATE","body": {"@type": "SSAPBodyUpdateMessage","ontology": '
			+ ontology + ',"query": "'
			+ query + '"}}';

		return queryUpdate;
	},

	/**
	 * UPDATE_BY_ID Operation
	 */
	generateUpdateByIdMessage : function(id, data, ontology, sessionKey) {
		if(ontology===undefined || ontology==null){
			ontology='null'
		}else{
			ontology='"'+ontology+'"';
		}

		var queryUpdate = '{"messageId": null,"sessionKey":"'
			+ sessionKey + '","direction": "REQUEST","messageType": "UPDATE_BY_ID","body": {"@type": "SSAPBodyUpdateByIdMessage","ontology": '
			+ ontology + ', "id":"'
			+ id + '","data":'
			+ data + '}}';

		return queryUpdate;
	}


}

exports.generateJoinByTokenMessage = SSAPMessageGenerator.generateJoinByTokenMessage;
exports.generateJoinRenovateByTokenMessage = SSAPMessageGenerator.generateJoinRenovateByTokenMessage;
exports.generateLeaveMessage = SSAPMessageGenerator.generateLeaveMessage;
exports.generateInsertMessage = SSAPMessageGenerator.generateInsertMessage;
exports.generateQueryWithQueryTypeMessage = SSAPMessageGenerator.generateQueryWithQueryTypeMessage;
exports.generateDeleteMessage = SSAPMessageGenerator.generateDeleteMessage;
exports.generateDeleteByIdMessage = SSAPMessageGenerator.generateDeleteByIdMessage;
exports.generateUpdateMessage = SSAPMessageGenerator.generateUpdateMessage;
exports.generateUpdateByIdMessage = SSAPMessageGenerator.generateUpdateByIdMessage;
